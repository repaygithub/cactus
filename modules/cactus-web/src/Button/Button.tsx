import styled, { FlattenInterpolation, ThemeProps, css } from 'styled-components'
import { space, SpaceProps } from 'styled-system'
import { CactusTheme } from '@repay/cactus-theme'

export type ButtonVariants = 'standard' | 'action'

interface ButtonProps extends SpaceProps {
  variant?: ButtonVariants
  disabled?: boolean
}

type VariantMap = { [K in ButtonVariants]: FlattenInterpolation<ThemeProps<CactusTheme>> } & {
  disabled: FlattenInterpolation<ThemeProps<CactusTheme>>
}

const variantMap: VariantMap = {
  action: css`
    color: ${p => p.theme.colors.callToActionText};
    background-color: ${p => p.theme.colors.callToAction};
    border-color: ${p => p.theme.colors.callToAction};

    &:hover,
    &:focus {
      color: ${p => p.theme.colors.baseText};
      background-color: ${p => p.theme.colors.base};
      border-color: ${p => p.theme.colors.base};
    }
  `,
  standard: css`
    color: ${p => p.theme.colors.base};
    background-color: ${p => p.theme.colors.white};
    border-color: ${p => p.theme.colors.base};

    &:hover,
    &:focus {
      color: ${p => p.theme.colors.baseText};
      background-color: ${p => p.theme.colors.base};
      border-color: ${p => p.theme.colors.base};
    }
  `,
  disabled: css`
    color: ${p => p.theme.colors.mediumGray};
    background-color: ${p => p.theme.colors.lightGray};
    border-color: ${p => p.theme.colors.lightGray};
    cursor: not-allowed;
  `,
}

const variantOrDiabled = (
  props: ButtonProps
): FlattenInterpolation<ThemeProps<CactusTheme>> | undefined => {
  if (props.disabled) {
    return variantMap.disabled
  } else if (props.variant !== undefined) {
    return variantMap[props.variant]
  }
}

export const Button = styled.button<ButtonProps>`
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  font-size: 18px;
  line-height: 1.5em;
  ${space}
  ${variantOrDiabled}
`

Button.defaultProps = {
  variant: 'standard',
  disabled: false,
  type: 'button',
}

export default Button
