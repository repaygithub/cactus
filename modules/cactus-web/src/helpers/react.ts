import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
