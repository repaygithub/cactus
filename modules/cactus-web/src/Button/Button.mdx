---
name: Button
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import Button from './Button'
import cactusTheme from '@repay/cactus-theme'
import Link from '../Link/Link'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# Button

### Try it out

export const code = `<Button variant='action' margin={20}>Take Some Action</Button>`

<LiveProvider code={code} scope={{ Button, Link }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Best practices

- `Button` components typically should not be used for navigation between pages.
- If you want to make a link look like a button, there are a couple of solutions, depending on whether or not you are using Typescript. However, this will modify the button's styles.
  - JSX Solution:
  ```jsx
  <Button as={Link} to="/path/to/something">
    Go Somewhere!
  </Button>
  ```
  - TSX Solution:
  ```tsx
  const ButtonLink = Button.withComponent(Link)
  <ButtonLink to='/path/to/something'>Go Somewhere!</ButtonLink>
  ```

## Basic usage

```jsx
import { Button } from '@repay/cactus-web'
...
// Basic usage
<Button>
  Basic Button
</Button>

// Call to action variant
<Button variant="action">
  Take Some Action
</Button>

// Inverse
<Button inverse>
  Inversed
</Button>

// Disabled
<Button disabled>
  Disabled
</Button>
```

## Properties

<PropsTable of={Button} />
