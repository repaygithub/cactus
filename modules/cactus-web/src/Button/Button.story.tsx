import React from 'react'
import { storiesOf } from '@storybook/react'
import { text, boolean, select } from '@storybook/addon-knobs/react'
import { actions } from '@storybook/addon-actions'
import Button, { ButtonVariants } from './Button'

const buttonVariants: ButtonVariants[] = ['standard', 'action']
const eventLoggers = actions('onClick', 'onFocus', 'onBlur')

storiesOf('Button', module).add(
  'Basic Usage',
  () => (
    <Button
      variant={select('variant', buttonVariants, 'standard')}
      disabled={boolean('disabled', false)}
      {...eventLoggers}
    >
      {text('children', 'A Button')}
    </Button>
  ),
  { options: { showPanel: true } }
)
