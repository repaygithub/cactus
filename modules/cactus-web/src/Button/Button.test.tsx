import * as React from 'react'
import { cleanup, render, fireEvent } from 'react-testing-library'
import Button from './Button'
import { ThemeProvider } from 'styled-components'
import cactusTheme from '@repay/cactus-theme'

afterEach(cleanup)

describe('component: Button', () => {
  test('should default to standard variant', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button>Click me!</Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should render standard variant', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button variant="standard">Click me!</Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should render call to action variant', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button variant="action">Click me!</Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should render disabled variant', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button disabled>Click me!</Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should render inverse standard variant', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button variant="standard" inverse>
          Click me!
        </Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should render inverse call to action variant', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button variant="action" inverse>
          Click me!
        </Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should render inverse disabled variant', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button disabled inverse>
          }>Click me!
        </Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should support margin space props', () => {
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button mt={5}>I have margins!</Button>
      </ThemeProvider>
    )

    expect(button.asFragment()).toMatchSnapshot()
  })

  test('should not support padding space props', () => {
    const _error = console.error
    console.error = jest.fn()
    const button = render(
      <ThemeProvider theme={cactusTheme}>
        <Button pt={5} />
      </ThemeProvider>
    )

    expect(console.error).toHaveBeenCalledWith(
      'Padding props are not supported! The value 5 for prop pt will have no effect on Button.'
    )
    expect(button.asFragment()).toMatchSnapshot()
    console.error = _error
  })

  test('should trigger onClick', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(
      <ThemeProvider theme={cactusTheme}>
        <Button onClick={onClick} variant="action" data-testid="clicked">
          Click me!
        </Button>
      </ThemeProvider>
    )

    fireEvent.click(getByTestId('clicked'))
    expect(onClick).toHaveBeenCalled()
  })

  test('should not trigger disabled onClick', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(
      <ThemeProvider theme={cactusTheme}>
        <Button onClick={onClick} variant="action" disabled data-testid="not-clicked">
          Click me!
        </Button>
      </ThemeProvider>
    )

    fireEvent.click(getByTestId('not-clicked'))
    expect(onClick).not.toHaveBeenCalled()
  })
})
