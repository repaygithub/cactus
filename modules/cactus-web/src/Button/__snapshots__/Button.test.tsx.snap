// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Button should default to standard variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(200,96%,11%);
  background-color: hsl(0,0%,100%);
  border-color: hsl(200,96%,11%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,11%);
  border-color: hsl(200,96%,11%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      Click me!
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render Spinner when loading is true 1`] = `
<DocumentFragment>
  .c2 {
  font-size: 16px;
  vertical-align: middle;
  -webkit-animation: bKPVWD 0.75s linear infinite;
  animation: bKPVWD 0.75s linear infinite;
}

.c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(200,96%,11%);
  background-color: hsl(0,0%,100%);
  border-color: hsl(200,96%,11%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .c1 {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,11%);
  border-color: hsl(200,96%,11%);
}

<button
    aria-live="assertive"
    class="c0"
    disabled=""
    type="button"
  >
    <span
      aria-hidden="true"
      style="visibility: hidden;"
    >
      Submit
    </span>
    <svg
      aria-label="loading"
      class="sc-qYiqT c1 c2"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M12,0 C5.372583,0 0,5.372583 0,12 C0,18.627417 5.372583,24 12,24 C18.627417,24 24,18.627417 24,12 C24,11.4477153 23.5522847,11 23,11 C22.4477153,11 22,11.4477153 22,12 C22,17.5228475 17.5228475,22 12,22 C6.4771525,22 2,17.5228475 2,12 C2,6.4771525 6.4771525,2 12,2 C12.5522847,2 13,1.55228475 13,1 C13,0.44771525 12.5522847,0 12,0 Z"
        transform="rotate(-177 12 12)"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render call to action variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,35%);
  border-color: hsl(200,96%,35%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,11%);
  border-color: hsl(200,96%,11%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      Click me!
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render danger variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(0,0%,100%);
  background-color: hsl(353,84%,44%);
  border-color: hsl(353,84%,44%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(0,0%,100%);
  background-color: hsl(353,96%,11%);
  border-color: hsl(353,96%,11%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      I am dangerous
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render disabled variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(0,0%,70%);
  background-color: hsl(0,0%,90%);
  border-color: hsl(0,0%,90%);
  cursor: not-allowed;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

<button
    aria-live="assertive"
    class="c0"
    disabled=""
    type="button"
  >
    <span>
      Click me!
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render inverse call to action variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,35%);
  border-color: hsl(200,96%,35%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(200,96%,35%);
  background-color: hsl(0,0%,100%);
  border-color: hsl(0,0%,100%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      Click me!
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render inverse danger variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(353,84%,44%);
  background-color: hsl(0,0%,100%);
  border-color: hsl(353,84%,44%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(0,0%,100%);
  background-color: hsl(353,84%,44%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      I am also dangerous
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render inverse disabled variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(0,0%,70%);
  background-color: hsl(0,0%,90%);
  border-color: hsl(0,0%,90%);
  cursor: not-allowed;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

<button
    aria-live="assertive"
    class="c0"
    disabled=""
    type="button"
  >
    <span>
      }&gt;Click me!
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render inverse standard variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,11%);
  border-color: hsl(0,0%,100%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(200,96%,11%);
  background-color: hsl(0,0%,100%);
  border-color: hsl(0,0%,100%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      Click me!
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should render standard variant 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  color: hsl(200,96%,11%);
  background-color: hsl(0,0%,100%);
  border-color: hsl(200,96%,11%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,11%);
  border-color: hsl(200,96%,11%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      Click me!
    </span>
  </button>
</DocumentFragment>
`;

exports[`component: Button should support margin space props 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  border-radius: 20px;
  padding: 2px 30px;
  border: 2px solid;
  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  font-size: 18px;
  line-height: 28px;
  margin-top: 24px;
  color: hsl(200,96%,11%);
  background-color: hsl(0,0%,100%);
  border-color: hsl(200,96%,11%);
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 10px);
  width: calc(100% + 10px);
  top: -5px;
  left: -5px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
  box-sizing: border-box;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
}

.c0 .sc-psCJM {
  position: absolute;
  left: 50%;
  top: 50%;
  margin-left: -8px;
  margin-top: -8px;
}

.c0:hover {
  color: hsl(0,0%,100%);
  background-color: hsl(200,96%,11%);
  border-color: hsl(200,96%,11%);
}

<button
    aria-live="assertive"
    class="c0"
    type="button"
  >
    <span>
      I have margins!
    </span>
  </button>
</DocumentFragment>
`;
